File: muni.hs
import Text.XML.Light
import Network.HTTP
import Network.Mail.Mime
import qualified Data.ByteString.Lazy.Char8 as B
import System.Environment

main = do
  (recip:needstr:sid:rt:extra) <- getArgs
  --let recip = "rjsen@me.com" --"4082300301@txt.att.net"
  --let needed = 15
  --let sid = "16619"
  --let rt = "L"
  let needed = read needstr :: Int

  res <- simpleHTTP $ getRequest("http://webservices.nextbus.com/service/publicXMLFeed?command=predictions&a=sf-muni&stopId=" ++ sid ++ "&routeTag=" ++ rt)
    
  xml <- getResponseBody res

  let Just min = parseXMLDoc xml >>= \x -> findAttr (QName "minutes" Nothing Nothing ) (head $ findElements (QName "prediction" Nothing Nothing ) x)
  let minutes = read min :: Int
  
  let msg = B.pack $ "To: " ++ recip ++ "\nThe next " ++ rt ++ " will be at stop " ++ sid ++ " in " ++ min ++ " minutes"

  if needed >= minutes
     then sendmail msg
     else return ()

File: muni.pl
#!/usr/bin/perl

use LWP::UserAgent;
use XML::Simple;
use Email::Simple;
use Email::Sender::Simple qw(sendmail);
use Getopt::Long;
use Data::Dumper;

my ($recip, $needed, $sid, $rt);
GetOptions ('recip=s' => \$recip,
	    'needed=i' => \$needed,
	    'sid=i' => \$sid,
	    'rt=s' => \$rt) || die 'not enough args';

my $ua = LWP::UserAgent->new;
my $res = $ua->get("http://webservices.nextbus.com/service/publicXMLFeed?command=predictions&a=sf-muni&stopId=".$sid."&routeTag=".$rt);
if ($res->is_success) {
  my $doc = XMLin($res->content, ForceArray => ['direction']);

  my $min = $doc->{predictions}->{direction}->[0]->{prediction}->[0]->{minutes};
  if ($needed >= $min) {
    my $email = Email::Simple->create(
				      header => [
						 From => 'rjsen@me.com',
						 To => $recip
						],
				      body => "The next $rt will be at stop $sid in $min minutes"
				     );
    eval {
      sendmail($email);
    };
    die "email failed: $@" if $@;
  }
} else {
  print "request failed: ".$res->status_line."\n";
}
